embassy-rp/src/clocks.rs

```diff
@@ -2,10 +2,139 @@ use pac::clocks::vals::*;
+    peri_clk_src: Option<ClkPeriCtrlAuxsrc>,
+    usb_clk_config: Option<(ClkUsbCtrlAuxsrc, u8)>,
+    adc_clk_config: Option<(ClkAdcCtrlAuxsrc, u8)>,
+    rtc_clk_config: Option<(ClkRtcCtrlAuxsrc, u32)>,
+}
+
+impl ClockConfig {
+    pub fn crystal(crystal_hz: u32) -> Self {
+        Self {
+            rosc_config: Some(RoscConfig {
+                range: pac::rosc::vals::FreqRange::MEDIUM,
+                drive_strength_0: 0,
+                drive_strength_1: 0,
+                drive_strength_2: 0,
+                drive_strength_3: 0,
+                drive_strength_4: 0,
+                drive_strength_5: 0,
+                drive_strength_6: 0,
+                drive_strength_7: 0,
+                div: 16,
+            }),
+            xosc_config: Some(XoscConfig {
+                hz: crystal_hz,
+                clock_type: ExternalClock::Crystal,
+                sys_pll: Some(PllConfig {
+                    refdiv: 1,
+                    vco_freq: 1500_000_000,
+                    post_div1: 6,
+                    post_div2: 2,
+                }),
+                usb_pll: Some(PllConfig {
+                    refdiv: 1,
+                    vco_freq: 480_000_000,
+                    post_div1: 5,
+                    post_div2: 2,
+                }),
+            }),
+            ref_clk_config: (RefClkSrc::Xosc, 1),
+            sys_clk_config: (SysClkSrc::Aux(ClkSysCtrlAuxsrc::CLKSRC_PLL_SYS), 1),
+            peri_clk_src: Some(ClkPeriCtrlAuxsrc::CLK_SYS),
+            usb_clk_config: Some((ClkUsbCtrlAuxsrc::CLKSRC_PLL_USB, 1)),
+            adc_clk_config: Some((ClkAdcCtrlAuxsrc::CLKSRC_PLL_USB, 1)),
+            rtc_clk_config: Some((ClkRtcCtrlAuxsrc::CLKSRC_PLL_USB, 1024)),
+        }
+    }
+
+    pub fn rosc() -> Self {
+        Self {
+            rosc_config: Some(RoscConfig {
+                range: pac::rosc::vals::FreqRange::HIGH,
+                drive_strength_0: 0,
+                drive_strength_1: 0,
+                drive_strength_2: 0,
+                drive_strength_3: 0,
+                drive_strength_4: 0,
+                drive_strength_5: 0,
+                drive_strength_6: 0,
+                drive_strength_7: 0,
+                div: 1,
+            }),
+            xosc_config: None,
+            ref_clk_config: (RefClkSrc::Rosc, 4),
+            sys_clk_config: (SysClkSrc::Aux(ClkSysCtrlAuxsrc::ROSC_CLKSRC), 1),
+            peri_clk_src: Some(ClkPeriCtrlAuxsrc::ROSC_CLKSRC_PH),
+            usb_clk_config: None,
+            adc_clk_config: Some((ClkAdcCtrlAuxsrc::ROSC_CLKSRC_PH, 1)),
+            rtc_clk_config: Some((ClkRtcCtrlAuxsrc::ROSC_CLKSRC_PH, 1024)),
+        }
+    }
+}
+pub enum ExternalClock {
+    Crystal,
+    Clock,
+}
+
+pub struct XoscConfig {
+    hz: u32,
+    clock_type: ExternalClock,
+    sys_pll: Option<PllConfig>,
+    usb_pll: Option<PllConfig>,
+}
+
+pub struct RoscConfig {
+    range: pac::rosc::vals::FreqRange,
+    drive_strength_0: u8,
```

Comment:
perhaps it'd make sense to make this an array? 

(disclaimer: I don't know what it is / what it's used for, lol)

Reply:
Doh' of course this should be an array. 

The drive strength is the input current for each stage of the ring oscillator. Basically bigger number = higher freq. Honestly I doubt anyone will ever use it. When the rosc is set to the high freq range it's already pushing the max freq of the core.